README.md

```diff
@@ -1,86 +1,108 @@
+
+Furthermore some third party vendors want to integrate the YAP search with their own applications. They want to create an image query and then get notified when any new images matching the query are uploaded. However each vendor have various limitations on the amount and rate of traffic they can handle, so YAP builds in rate limits and circuit breakers patterns for the different vendors. In this complex architecture, YAP finds it **difficult to reason about performance bottlenecks and when and where they can occur in the application**, especially when adding new integrations.
+
+Furthermore now YAP has become a multi-tenant search engine while itself running on top a multi-tenant infrastructure provider. YAP is facing considerable difficulty understanding the cost of onboarding new tenants, **modelling the cost of building new features** across tenants. This is further complicated cloud providers have granular charges (upto per request or per GB consumed cost) and tenants have diverse traffic patterns.
+
+### Idea
+
+The three problems mentioned above, namely reasoning about performance, enforcing invariants and reasing about cost are inter-related and point to the lack of analysis tools that reason across cloud components. Most of these properties can only be determined by through logs, traces and metrics generated by running the system which is costly and error prone.
+
+A modern programming system should do allow the following -
+
+* Declare logical invariants for the data flow specification. Simluate the application as a state machine to check if invariants are satified. Derive the working implementation from the specification [^3](#3-protocol-combinators) to ensure there is no gap.
+* Perform data flow analysis and consider performance characteristics of various architectural components to identify bottlenecks.
+* Use cost models of various architectural components and sample data traces or probabilistic traffic patterns to derive cost estimates without deploying/running the application.
+
+## Dynamic application logic based on infrastructure state
+
+### Story
+
+YAP has found a clever way to deal with peak traffic without spending etc money. YAP dynamically changes their query logic accuracy based on the traffic they are receiving. When they have auto-scaled to 8 query server instances and have reached the 70% CPU utilization threshold they drop search accuracy from 95% to 90%. They cannot compromise on end user latency, but the 5% drop in accuracy is barely noticed so they use this technique to handle brief bursty traffic without incurring extra cost.
+
+### Idea
+
+The underlying runtime state needs to expose properties to the application logic. Then the application logic can change it's behaviour based on the state. Coupled with some ideas mentioned above like checking invariants and static analysis this approach will make it possible to truly utilize the cloud system as an extension of the application.
```

Comment:
While the idea itself is pretty awesome, you are basically creating a non deterministic system, i think there is an element of coupling decoupling here that i feel makes simple systems more complicated. Basically the result behavior is not easily unit testable or integration testsable easily..

Reply:
Aah yes this needs to be in an ambitious or oddball section, and a bit more fleshed out. But I still feel it should be there especially when dynamodb itself utilizes feedback mechanisms. Having a platform expose such functionality would be interesting to see.

Adding relevant dynamodb youtube link - https://youtu.be/yvBR71D0nAQ?t=1179